### ISSUE TITLE
[iOS] Safari is using memory cache instead of service worker

### ISSUE THREAD

**Issue by kawazoe** [GoogleChrome/workbox#1744]
## Library Affected
workbox-sw 3.6.2

## Browser & Platform
iOS 12.0.1 through 12.1.4; Safari home-screen app. Did not test 12.2 yet.

## Issue
**Context**
 - vuejs application
   - created with cli v3.0.5
   - PWA enabled
   - all other options defaults
 - hosted on Azure Web Apps (plain IIS static site)
 - using attached web.config
 - works as expected in Chrome 70 on Windows 10

I chose to post this issue here rather than on the vuejs repository since there doesn't appear to be any special behavior added by vue in such cases. This app uses a basic, auto-generated SW with a precache manifest listing the build files and files in the public folder. The app scripts are injected in the index.html at build time and have versioned (hashed) filenames. The only changes are the `web.config` file below and a sentence change on the home page to test cache busting.

**Initial state**
When loading from the home-screen for the first time, the application appears to works as expected. A service worker is registered and the app starts in airplane mode (offline).

**Trigger**
Push an updated version of the app (bundles has different hashes) on the server. For instance, change a sentence on the home page.

**Expected**
- Restart* the app
  - A new SW is detected and gets installed
  - The new SW builds its offline cache with the new code
- Restart the app
  - The new SW take over and serve the new version of the app

**Actual**
- Restart the app
  - A new SW is detected and gets installed
- Restart the app
  - The old app gets loaded

**Additional testing after the second restart**
- Use the Mac OS Safari DevTools
  - Refresh (not forced) the page
    - Notice that the bundles are served from *Memory*
  - Force refresh the page
    - All requests goes through to the web server
  - Refresh the page
    - All requests served from *Service worker* (as expected)
  - Refresh the page again
    - Notice that the bundles are served from *Memory* again with the old code

There is also a chance that doing the refresh operation will corrupt the SW cache and thus serving a mix of the old and new code (typically the new index.html and old code bundles, but refreshed straight from the server). In this state, the offline mode becomes completely unusable.

*Restart meaning either closing the tab in Chrome/Win10 or using the app switcher and swiping up on iOS/safari.

## Notes
I have verified that all caching headers made it to the client device properly but this doesn't mean that this issue isn't related to cache-control. If you see something invalid in my cache-control setup, I will retest and update this issue accordingly.

## web.config
```
<configuration>
    <system.webServer>
        <httpProtocol>
            <customHeaders>
                <add name="Pragma" value="no-cache" />
                <add name="Expires" value="0" />
            </customHeaders>
        </httpProtocol>
        <staticContent>
            <clientCache cacheControlCustom="no-cache, no-store, must-revalidate, max-age=0" setEtag="false" />
            <mimeMap fileExtension=".json" mimeType="application/json" />
        </staticContent>
        <rewrite>
            <rewriteMaps>
                <rewriteMap name="^(.*)$" />
            </rewriteMaps>
            <rules>
                <rule name="Serve root on missing file (SPA routing)" stopProcessing="true">
                    <match url="^(.*)$" />
                    <conditions logicalGrouping="MatchAll">
                        <add input="{REQUEST_FILENAME}"
                             matchType="IsFile"
                             negate="true" />
                    </conditions>
                    <action type="Rewrite" url="/" />
                </rule>
            </rules>
        </rewrite>
    </system.webServer>
</configuration>
```

**Comment by tdsereno** [comment]
Versão 4.0 – Utilização obrigatória a partir de 01/10/2018:
http://nfe.sefaz.ba.gov.br/servicos/nfce/qrcode.aspx (deverá compor o QR Code para
acesso ao endereço de consulta)
Após a página, deve ser inserido o caractere “?” e os parâmetros."

Fonte:
https://nfe.sefaz.ba.gov.br/servicos/nfce/documentos/configuracao_programa_emissor_REV_3.pdf

**Comment by kawazoe** [comment]
I will wait for the fix to land in iOS and check again later. I am not sure this is actually the same issue, but it might be the same fix. If the issue is still there after 1902569 gets fixed, I'll open a separate issue in webkit for this.

**Comment by mikeli11** [comment]
@kawazoe I fixed my issue with Safari not updating to the latest SW by adding the skipwaiting:true option in the vue.config.js. After restarting the app on iOS, the latest version is now loaded.

workboxOptions: { skipWaiting: true }

Let me know if this helps you with this issue. I don't know if this is the expected behaviour in Safari since this setting force installs the latest SW as soon as it's downloaded (which is not recommended). As you know already, chrome doesn't need this for the expected behaviour.

Reference:
https://developers.google.com/web/tools/workbox/modules/workbox-webpack-plugin

**Comment by kawazoe** [comment]
@mikejamesli This is interesting. Have you tried only once or multiple updates? What I have experienced is that usually, the first update will go through, but then Safari returns a cached page before the service worker intercepts it.

**Comment by mikejamesli** [comment]
@kawazoe If what you mean by multiple updates is updating to a version of the service worker to one several versions higher than the current one then yes, it does update on restart.

**Comment by mikejamesli** [comment]
@kawazoe i'm glad that helped you.

Unfortunately this isn't the ideal solution for me as I've since witnessed the "Lazy-loading of fingerprinted content" issue found in this [stackoverflow post](https://stackoverflow.com/questions/51715127/what-are-the-downsides-to-using-skipwaiting-and-clientsclaim-with-workbox). However, it is the only way I can update my service worker on iOS devices (iPad 4 mini) at the moment.

**Comment by jeffposnick** [comment]
Just wanted to ping this issue to see if folks have been able to reproduce it using the most up to date iOS/Safari release.

**Comment by kawazoe** [comment]
I did not get a chance to try it out the newer versions of Safari without @mikejamesli 's workaround yet. I will probably spend some time on this in the next two weeks though so expect some updates from me around then.

**Comment by TP18** [comment]
Invalid options in vue.config.js: "workboxOptions" is not allowed

**Comment by TP18** [comment]
The safari version on iOS 12.2 still caches the old content.

**Comment by kawazoe** [comment]
@TP18 The setting is not at the root of the config. It's under 'pwa'. It should look something like this:

```js
module.exports = {
  ...,
  pwa: {
    workboxOptions: {
      skipWaiting: true,
    },
  },
};
```

**Comment by hasselboxster** [comment]
On iOS Safari - i have found that if you "Request Desktop Site" you will get the new version.  Its a quick work around, but the average user/consumer of our app is not going to know or want to do this each time a new version of our app is released.

**Comment by marlonmleite** [comment]
I have the same problem and have already used more than hours here in the company to solve the same, but without success.

We made several versions of our app, with automatic skipWaiting and also with skipWaiting being called by the user. All of our solutions work on just about anything, however in iOS PWA not work.

Works well on Chrome, Firefox, Android, Safari. However, in PWA on iOS it does not work at all (we use more than 5 different implementations and none at all, just restarting the device).

Has anyone evolved from this?

**Comment by kawazoe** [comment]
I can confirm that the issue I was having, caused by the service worker cache not being shared between safari and an home screen app, has been fixed in iOS 12.2. If you are having a different issue, such as your app not updating, then please look in the thread for the skipWaiting workaround.

**Comment by marlonmleite** [comment]
> I can confirm that the issue I was having, caused by the service worker cache not being shared between safari and an home screen app, has [been fixed in iOS 12.2](https://bugs.webkit.org/show_bug.cgi?id=190269). If you are having a different issue, such as your app not updating, then please look in the thread for the skipWaiting workaround. You will also want to look at https://gist.github.com/kawazoe/fa3b5a3c998d16871ffb9e2fd721cb4b which irons out some of the issues I have seen with iOS. Keep in mind that this is a hack/workaround and should be taken with a grain of salt.

I opened this bug for the problem: https://bugs.webkit.org/show_bug.cgi?id=199110

**Comment by jcrooke** [comment]
Not sure if anyone will see this but if you do happen to stumble across this issue (like I did), trying to figure out why in the world your PWA doesn't work on iOS when it works perfectly on Desktop - I discovered it's because the max cache "limit" that is reported to be 50MB, is actually more like 10MB.  If you exceed the limit, your PWA just won't work offline.  My solution was to only cache necessary items and to compress the hell out of images and audio.
